File: code/packages/java/wasm-types/src/test/java/com/codingadventures/wasmtypes/WasmTypesTest.java
package com.codingadventures.wasmtypes;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

class WasmTypesTest {
    @Test
    void buildsFunctionTypesAndModuleContainers() {
        WasmTypes.FuncType signature = WasmTypes.makeFuncType(
                List.of(WasmTypes.ValueType.I32),
                List.of(WasmTypes.ValueType.I64)
        );

        WasmModule module = new WasmModule();
        module.types.add(signature);
        module.functions.add(0);

        assertEquals(1, module.types.size());
        assertEquals(WasmTypes.ValueType.I32, signature.params().get(0));
        assertEquals(WasmTypes.ValueType.I64, signature.results().get(0));
        assertNull(module.start);
    }

    @Test
    void preservesBytePayloadsInStructuralTypes() {
        WasmTypes.Global global = new WasmTypes.Global(
                new WasmTypes.GlobalType(WasmTypes.ValueType.I32, false),
                new byte[]{0x41, 0x2A, 0x0B}
        );
        WasmTypes.DataSegment segment = new WasmTypes.DataSegment(0, new byte[]{0x41, 0x00, 0x0B}, new byte[]{1, 2, 3});

        assertEquals(3, global.initExpr().length);
        assertEquals(3, segment.data().length);
    }
}

File: code/packages/kotlin/wasm-types/src/test/kotlin/com/codingadventures/wasmtypes/WasmTypesTest.kt
package com.codingadventures.wasmtypes

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class WasmTypesTest {
    @Test
    fun buildsFunctionTypesAndModuleContainers() {
        val signature = makeFuncType(listOf(ValueType.I32), listOf(ValueType.I64))
        val module = WasmModule()
        module.types += signature
        module.functions += 0

        assertEquals(1, module.types.size)
        assertEquals(ValueType.I32, signature.params.first())
        assertEquals(ValueType.I64, signature.results.first())
        assertNull(module.start)
    }

    @Test
    fun preservesBytePayloadsInStructuralTypes() {
        val global = Global(GlobalType(ValueType.I32, false), byteArrayOf(0x41, 0x2A, 0x0B))
        val segment = DataSegment(0, byteArrayOf(0x41, 0x00, 0x0B), byteArrayOf(1, 2, 3))

        assertEquals(3, global.initExpr.size)
        assertEquals(3, segment.data.size)
    }
}
